Update RELEASES.md for 1.55.0 (#87624)
rust-lang/rust: RELEASES.md

Version 1.55.0 (2021-09-09)
============================

Language
--------
- [You can create bindings inside of a `@` bindings pattern.][85305] E.g. the
  following is now possible.
  ```rust
  let outer_option = Some(42);

  match outer_option {
      option @ Some(value) => {

>          // We can now use both `option` and `value`.

Author: Updated the example to work with any type.

         ...
      }
      _ => { ... }
  }
  ```

> - [You can now write open "from" ranges (`X..`), which will start at `X` and

Reviewer: ```suggestion
- [You can now write open "from" range patterns (`X..`), which will start at `X` and
```

  will end at the maximum value of the integer.][83918]
- [You can now access `union` fields in `const` contexts.][85769]
- [You can now explicitly import the prelude of different editions

>   through `std::prelude`. (e.g. `use std::prelude::rust_2021::*;`)][86294]

Reviewer 2: ```suggestion
  through `std::prelude` (e.g. `use std::prelude::rust_2021::*;`).][86294]
```

Compiler
--------
- [Added tier 3\* support for `powerpc-unknown-freebsd`.][87370]
- [Added tier 3 support for `powerpc64le-unknown-freebsd`.][83572]

\* Refer to Rust's [platform support page][platform-support-doc] for more
   information on Rust's tiered platform support.

Libraries
---------

- [Updated std's float parsing to use the Eisel-Lemire algorithm.][86761]
  These improvements should in general provide faster string parsing of floats,
  increases the range of accepted valid floating point values, and reduce
  the produced code size for non-stripped artifacts.
- [`string::Drain` now implements `AsRef<str>` and `AsRef<[u8]>`.][86858]
- [`collections::{BinaryHeap, BTreeSet, HashSet, LinkedList, VecDeque}` now
  implement `From<[T; N]>`.][84111]

> - [`collections::{BTreeMap, HashMap}` now implement `From<[(K, V); N]>`.][84111]

Reviewer 2: Would it be worth mentioning how this allows for literal-like syntax, as [described in the PR](#84111)?

Stabilised APIs
---------------

- [`Bound::cloned`]
- [`Drain::as_str`]
- [`IntoInnerError::into_error`]
- [`IntoInnerError::into_parts`]
- [`MaybeUninit::assume_init_mut`]
- [`MaybeUninit::assume_init_ref`]
- [`MaybeUninit::write`]
- [`array::map`]
- [`ops::ControlFlow`]
- [`x86::_bittest`]
- [`x86::_bittestandcomplement`]
- [`x86::_bittestandreset`]
- [`x86::_bittestandset`]
- [`x86_64::_bittest64`]
- [`x86_64::_bittestandcomplement64`]
- [`x86_64::_bittestandreset64`]
- [`x86_64::_bittestandset64`]

The following previously stable functions are now `const`.

- [`str::from_utf8_unchecked`]
- [`mem::transmute`]

> Cargo

Reviewer: I just posted the release notes for cargo: https://github.com/rust-lang/cargo/pull/9748.  I never know what people might be interested in highlighting here.

One thing people mentioned wanting to highlight is https://github.com/rust-lang/cargo/pull/9675 (see https://github.com/rust-lang/cargo/pull/9675#issuecomment-881969102) and similarly https://github.com/rust-lang/cargo/pull/9655 (which together makes several changes to diagnostic output).

Reviewer 2: @ehuss Think it would be worth adding rust-lang/cargo#9680? It's a small addition, but adding it to the release notes might help discoverability. I can do a PR, if desired.

-----------------------
- [The package definition in `cargo metadata` now includes the `"default_run"`
  field from the manifest.][cargo/9550]

Compatibility Notes
-------------------
- [std functions that return an `io::Error` will no longer use the
  `ErrorKind::Other` variant.][85746] This is to better reflect that these
  kinds of errors could be categorised [into newer more specific `ErrorKind`
  variants][79965], and that they do not represent a user error.
- [Using environment variable names with `process::Command` on Windows now

>   behaves as expected][85270] Previously using envionment variables with

Reviewer 2: ```suggestion
  behaves as expected.][85270] Previously using envionment variables with
```

Version 1.55.0 (2021-09-09)
============================

Language
--------
- [You can create bindings inside of a `@` bindings pattern.][85305] E.g. the
  following is now possible.
  ```rust
  let outer_option = Some(42);

  match outer_option {
      option @ Some(ref value) => {
         // We can now use both `option` and `value`.
         ...
      }
      _ => { ... }
  }
  ```
- [You can now write open "from" range patterns (`X..`), which will start at `X` and
  will end at the maximum value of the integer.][83918]
- [You can now access `union` fields in `const` contexts.][85769]
- [You can now explicitly import the prelude of different editions
  through `std::prelude` (e.g. `use std::prelude::rust_2021::*;`).][86294]

Compiler
--------
- [Added tier 3\* support for `powerpc-unknown-freebsd`.][87370]
- [Added tier 3 support for `powerpc64le-unknown-freebsd`.][83572]

\* Refer to Rust's [platform support page][platform-support-doc] for more
   information on Rust's tiered platform support.

Libraries
---------

- [Updated std's float parsing to use the Eisel-Lemire algorithm.][86761]
  These improvements should in general provide faster string parsing of floats,

>   no longer rejects certain valid floating point values, and reduce

Reviewer 3: ```suggestion
  no longer reject certain valid floating point values, and reduce
```
This should probably say "reject" because "improvements" is plural.
